File: Data Structures & Algorithms/three-integer-sum/submission-4.java
class Solution {
  public List<List<Integer>> threeSum(int[] nums) {

    Arrays.sort(nums);
    List<List<Integer>> res = new ArrayList<>();
    Set<Integer> seen = new HashSet<>();
    for (int i = 0; i < nums.length - 2; i++) {
      if (seen.contains(nums[i])) {
        continue;
      }

      seen.add(nums[i]);

      int lPtr = i + 1;
      int rPtr = nums.length - 1;

      Set<Integer> seenInner = new HashSet<>();
      while (lPtr < rPtr) {
        int num1 = nums[i];
        int num2 = nums[lPtr];
        int num3 = nums[rPtr];

        int sum = num1 + num2 + num3;

        if (seenInner.contains(num3) || sum > 0) {
          rPtr--;
        } else if (seenInner.contains(num2) || sum < 0) {
          lPtr++;
        } else {
          res.add(List.of(num1, num2, num3));
          lPtr++;
          rPtr--;

          seenInner.add(num2);
          seenInner.add(num3);
        }
      }
    }
    return res;
  }


}

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-4.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        return self.binary_search(nums, target, 0, len(nums) -1)
        
    def binary_search(self, nums,target, lptr, rptr):
        if lptr == rptr:
            print(lptr)
            return lptr if nums[lptr] == target else -1
        
        mid = (rptr + lptr) // 2

        if nums[mid] == target:
            return mid

        print(lptr, rptr, mid)
        is_left_sorted = nums[mid] >= nums[lptr]

        if is_left_sorted:
            if target >= nums[lptr] and target < nums[mid]:
                return self.binary_search(nums,target, lptr, mid)
            else:
                return self.binary_search(nums,target, mid+1, rptr)
        else:
            if target > nums[mid] and target <= nums[rptr]:
                return self.binary_search(nums,target, mid+1, rptr)
            else:
                return self.binary_search(nums,target, lptr, mid)

        

